api/src/controllers/transferCredits.ts

```diff
@@ -111,54 +120,104 @@ const transferCreditsRouter = router({
+  }),
+  convertUserLegacyTransfers: publicProcedure.input(z.array(transferData)).query(async ({ input }) => {
+    return await organizeLegacyTransfers(input);
+  }),
+  overrideAllTransfers: userProcedure
+    .input(
+      z.object({
+        courses: z.array(zodCourseTransferSchema),
+        ap: z.array(zodAPExamSchema),
+        ge: z.array(zodGETransferSchema),
+        other: z.array(zodUncategorizedTransferSchema),
+      }),
+    )
     .mutation(async ({ ctx, input }) => {
-      const conditions = [eq(transferredMisc.userId, ctx.session.userId!)];
+      const appendUserId = <T extends object>(item: T) => Object.assign(item, { userId: ctx.session.userId! });
 
-      if (input.name != null) {
-        conditions.push(eq(transferredMisc.courseName, input.name));
-      } else {
-        conditions.push(isNull(transferredMisc.courseName));
+      const dbQueries = [];
+      if (input.courses.length) {
+        const addCoursesOperation = db
+          .insert(transferredCourse)
+          .values(input.courses.map(appendUserId))
+          .onConflictDoUpdate({
+            target: [transferredCourse.userId, transferredCourse.courseName],
+            set: { units: sql.raw(`EXCLUDED.${transferredCourse.units.name}`) },
+          });
+        dbQueries.push(addCoursesOperation);
       }
-
-      if (input.units != null) {
-        conditions.push(eq(transferredMisc.units, input.units));
-      } else {
-        conditions.push(isNull(transferredMisc.units));
+      if (input.ap.length) {
+        const addApQuery = db
+          .insert(transferredApExam)
+          .values(input.ap.map(appendUserId))
+          .onConflictDoUpdate({
+            target: [transferredApExam.userId, transferredApExam.examName],
+            set: {
+              score: sql.raw(`EXCLUDED.${transferredApExam.score.name}`),
+              units: sql.raw(`EXCLUDED.${transferredApExam.units.name}`),
+            },
+          });
+        dbQueries.push(addApQuery);
+      }
+      if (input.ge.length) {
+        const addGeQuery = db
+          .insert(transferredGe)
+          .values(input.ge.map(appendUserId))
+          .onConflictDoUpdate({
+            target: [transferredGe.userId, transferredGe.geName],
+            set: {
+              numberOfCourses: sql.raw(`EXCLUDED.${transferredGe.numberOfCourses.name}`),
+              units: sql.raw(`EXCLUDED.${transferredGe.units.name}`),
+            },
+          });
+        dbQueries.push(addGeQuery);
+      }
+      if (input.other.length) {
+        const rows = input.other.map(({ name: courseName, units }) => ({
+          courseName,
+          units,
+          userId: ctx.session.userId!,
+        }));
+        const addOtherQuery = db.insert(transferredMisc).values(rows);
+        dbQueries.push(addOtherQuery);
       }
 
-      await db.delete(transferredMisc).where(and(...conditions));
+      await Promise.all([dbQueries.map((q) => q.execute())]);
```

Comment:
Might be more appropriate to wrap this in a transaction since this mutation should be one logical step?

Reply:
Ideally yes, but I think this should not be an issue since we always update when there's conflict (and the schema prevents invalid values from entering in)